store/src/java/com/zimbra/cs/service/util/JWTUtil.java

```diff
@@ -0,0 +1,228 @@
+
+import java.security.Key;
+import java.util.Date;
+import java.util.Map;
+import java.util.UUID;
+
+import javax.crypto.spec.SecretKeySpec;
+import javax.servlet.http.HttpServletRequest;
+
+import com.google.common.primitives.Bytes;
+import com.zimbra.common.account.Key.AccountBy;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.soap.Element;
+import com.zimbra.common.soap.HeaderConstants;
+import com.zimbra.common.util.Constants;
+import com.zimbra.common.util.StringUtil;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.account.Account;
+import com.zimbra.cs.account.AccountServiceException;
+import com.zimbra.cs.account.AccountServiceException.AuthFailedServiceException;
+import com.zimbra.cs.account.AuthTokenKey;
+import com.zimbra.cs.account.JWTCache;
+import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.account.ZimbraJWT;
+import com.zimbra.soap.SoapEngine;
+import com.zimbra.soap.SoapServlet;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.ExpiredJwtException;
+import io.jsonwebtoken.JwtBuilder;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.MalformedJwtException;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.SignatureException;
+import io.jsonwebtoken.UnsupportedJwtException;
+
+public class JWTUtil {
+
+    /**
+     * Generate JWT based on input parameters
+     * @param jwtKey        final secret key used for signing jwt
+     * @param salt          random key associated with this jwt
+     * @param issuedAt      time at which jwt is issued
+     * @param expires       time after which jwt expires
+     * @param account       account for jwt is being generated
+     * @return              jwt
+     */
+    public static final String generateJWT(byte[] jwtKey, String salt, long issuedAt, long expires, Account account) {
+        Key key = new SecretKeySpec(jwtKey, SignatureAlgorithm.HS512.getJcaName());
+        JwtBuilder builder = Jwts.builder();
+        builder.setSubject(account.getId());
+        builder.setIssuedAt(new Date(issuedAt));
+        builder.setExpiration(new Date(expires));
+        String jti = UUID.randomUUID().toString();
+        builder.setId(jti);
+        builder.claim(Constants.TOKEN_VALIDITY_VALUE_CLAIM, account.getAuthTokenValidityValue());
+        JWTCache.put(jti, new ZimbraJWT(salt, expires));
+        return builder.signWith(SignatureAlgorithm.HS512, key).compact();
+    }
+
+    /**
+     * Look for the salt in request:soap context/soapEngine context/ZM_JWT cookie and return the value if found.
+     * @param soapCtxt
+     * @param engineCtxt
+     * @return
+     */
+    public static final String getSalt(Element soapCtxt, Map engineCtxt) {
+        String salt = null;
+        salt = soapCtxt == null ? null : soapCtxt.getAttribute(HeaderConstants.E_JWT_SALT, null);
+        if (salt == null && engineCtxt != null) {
+            ZimbraLog.account.debug("salt not found in soap context");
```

Comment:
Suggest deleting this and in the if branch, having:
`ZimbraLog.account.debug("salt not found in soap context. Using engine context salt=%s", salt);`